lib/recursive-methods.rb

```diff
@@ -1,49 +1,87 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  
+  return 1 if n <= 1
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(s.length) or O(n)
+# Space complexity: O(s.length) or O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  
+  return s[s.length-1] << reverse(s[1..-2]) << s[0]
 end
 
 # Time complexity: ?
-# Space complexity: ?
+# Space complexity: O(1)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  # How would I do this??
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+  
+  return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(s.length) or O(n)
+# Space complexity: O(1)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.nil?
+  return false if s.length == 1
+  
+  if s.length == 2
+    if s == "()"
+      return true
+    else
+      return false
+    end
+  end
+  
+  nested(s[1..-2])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(array.length) or O(n)
+# Space complexity: O(1)
 def search(array, value)
```

Comment:
👍 This works, but you have similar time/space issues with the above methods due to creating new arrays.